In [7]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/heartdisease/behavioral_risk_factor_data_heart_disease_stroke_prevention.csv


In [8]:
data = pd.read_csv("/kaggle/input/heartdisease/behavioral_risk_factor_data_heart_disease_stroke_prevention.csv", low_memory=False)

In [9]:
data.head()

year locationabbr locationdesc datasource priorityarea1  priorityarea2  \
0  2011           NV       Nevada      BRFSS           NaN            NaN   
1  2011           NV       Nevada      BRFSS           NaN            NaN   
2  2011           NV       Nevada      BRFSS           NaN            NaN   
3  2011           NV       Nevada      BRFSS           NaN            NaN   
4  2011           NV       Nevada      BRFSS           NaN            NaN   

  priorityarea3  priorityarea4                 category  \
0           NaN            NaN  Cardiovascular Diseases   
1           NaN            NaN  Cardiovascular Diseases   
2           NaN            NaN  Cardiovascular Diseases   
3           NaN            NaN  Cardiovascular Diseases   
4           NaN            NaN  Cardiovascular Diseases   

                    topic  ...  \
0  Coronary Heart Disease  ...   
1  Coronary Heart Disease  ...   
2  Coronary Heart Disease  ...   
3  Coronary Heart Disease  ...   
4  Coronary Heart Disease  ...   

                                 data_value_footnote confidence_limit_low  \
0  Statistically unstable estimates not presented...                  NaN   
1  Statistically unstable estimates not presented...                  NaN   
2  Statistically unstable estimates not presented...                  NaN   
3  Statistically unstable estimates not presented...                  NaN   
4  Statistically unstable estimates not presented...                  NaN   

  confidence_limit_high categoryid topicid  indicatorid breakoutcategoryid  \
0                   NaN          1      T4        BR001              BOC03   
1                   NaN          1      T4        BR001              BOC04   
2                   NaN          1      T4        BR001              BOC03   
3                   NaN          1      T4        BR001              BOC04   
4                   NaN          1      T4        BR001              BOC04   

  breakoutid  locationid                                    geolocation  
0      AGE04        32.0  POINT(-117.07184056399967 39.493240390000494)  
1      RAC02        32.0  POINT(-117.07184056399967 39.493240390000494)  
2      AGE01        32.0  POINT(-117.07184056399967 39.493240390000494)  
3      RAC07        32.0  POINT(-117.07184056399967 39.493240390000494)  
4      RAC04        32.0  POINT(-117.07184056399967 39.493240390000494)  

[5 rows x 27 columns]

In [10]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 35004 entries, 0 to 35003
Data columns (total 27 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   year                        35004 non-null  int64  
 1   locationabbr                35004 non-null  object 
 2   locationdesc                35004 non-null  object 
 3   datasource                  35004 non-null  object 
 4   priorityarea1               2704 non-null   object 
 5   priorityarea2               0 non-null      float64
 6   priorityarea3               11938 non-null  object 
 7   priorityarea4               0 non-null      float64
 8   category                    35004 non-null  object 
 9   topic                       35004 non-null  object 
 10  indicator                   35004 non-null  object 
 11  break_out_category          35004 non-null  object 
 12  break_out                   35004 non-null  object 
 13  data_value_type             350

In [11]:
data['confidence_limit_low'].unique()

array([ nan,  0.4,  1.5,  1.6,  0.3,  1.4,  0.8,  0.9,  1.9,  1.7,  2.3,
        0.5,  0.2,  0.1,  0.6,  0.7,  1. ,  1.1,  1.2,  1.3,  1.8,  2. ,
        2.1,  2.2,  2.4,  2.5,  2.6,  2.7,  2.8,  2.9,  3. ,  3.1,  3.3,
        3.2,  3.4,  3.5,  3.6,  3.7,  3.8,  3.9,  4. ,  4.1,  4.2,  4.3,
        4.4,  4.5,  4.6,  4.7,  4.9,  4.8,  5. ,  5.1,  5.4,  5.2,  5.3,
        5.5,  5.7,  5.6,  5.8,  6. ,  5.9,  6.1,  6.2,  6.3,  6.4,  6.5,
        6.6,  6.7,  6.8,  7. ,  6.9,  7.1,  7.4,  7.3,  7.2,  7.7,  7.5,
        7.9,  7.6,  7.8,  8. ,  8.1,  8.2,  8.3,  8.4,  8.5,  8.7,  8.6,
        8.9,  9. ,  8.8,  9.1,  9.5,  9.4,  9.2,  9.3,  9.7,  9.6,  9.8,
        9.9, 10. , 10.1, 10.3, 10.2, 10.4, 10.5, 10.6, 10.7, 10.8, 10.9,
       11. , 11.1, 11.2, 11.3, 11.4, 11.5, 12.1, 11.8, 11.6, 12. , 11.7,
       11.9, 12.3, 12.2, 12.4, 12.6, 12.7, 12.5, 12.8, 12.9, 13.1, 13. ,
       13.4, 13.2, 13.3, 13.6, 13.5, 13.9, 13.8, 14. , 13.7, 14.2, 14.3,
       14.5, 14.1, 14.4, 14.6, 14.7, 14.8, 14.9, 15

In [12]:
data.columns

Index(['year', 'locationabbr', 'locationdesc', 'datasource', 'priorityarea1',
       'priorityarea2', 'priorityarea3', 'priorityarea4', 'category', 'topic',
       'indicator', 'break_out_category', 'break_out', 'data_value_type',
       'data_value_unit', 'data_value', 'data_value_footnote_symbol',
       'data_value_footnote', 'confidence_limit_low', 'confidence_limit_high',
       'categoryid', 'topicid', 'indicatorid', 'breakoutcategoryid',
       'breakoutid', 'locationid', 'geolocation'],
      dtype='object')

## Import library

In [13]:
from sklearn.preprocessing import LabelEncoder
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from lightgbm import LGBMClassifier
from sklearn.model_selection import GridSearchCV
from sklearn.ensemble import RandomForestClassifier
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.metrics import accuracy_score, f1_score, mean_absolute_error, mean_squared_error

In [14]:
columns = ['year', 'category', 'topic', 'indicator', 'data_value','confidence_limit_low','confidence_limit_high']

## Lọc các cột cho tập dữ liệu

In [15]:
data_train =data[columns]
data_train

year                 category                   topic  \
0      2011  Cardiovascular Diseases  Coronary Heart Disease   
1      2011  Cardiovascular Diseases  Coronary Heart Disease   
2      2011  Cardiovascular Diseases  Coronary Heart Disease   
3      2011  Cardiovascular Diseases  Coronary Heart Disease   
4      2011  Cardiovascular Diseases  Coronary Heart Disease   
...     ...                      ...                     ...   
34999  2011  Cardiovascular Diseases  Coronary Heart Disease   
35000  2011  Cardiovascular Diseases  Coronary Heart Disease   
35001  2011  Cardiovascular Diseases  Coronary Heart Disease   
35002  2011  Cardiovascular Diseases  Coronary Heart Disease   
35003  2011  Cardiovascular Diseases  Coronary Heart Disease   

                                               indicator  data_value  \
0      Prevalence of coronary heart disease among US ...         NaN   
1      Prevalence of coronary heart disease among US ...         NaN   
2      Prevalence of coronary heart disease among US ...         NaN   
3      Prevalence of coronary heart disease among US ...         NaN   
4      Prevalence of coronary heart disease among US ...         NaN   
...                                                  ...         ...   
34999  Prevalence of coronary heart disease among US ...         NaN   
35000  Prevalence of coronary heart disease among US ...         NaN   
35001  Prevalence of coronary heart disease among US ...         NaN   
35002  Prevalence of coronary heart disease among US ...         NaN   
35003  Prevalence of coronary heart disease among US ...         NaN   

       confidence_limit_low  confidence_limit_high  
0                       NaN                    NaN  
1                       NaN                    NaN  
2                       NaN                    NaN  
3                       NaN                    NaN  
4                       NaN                    NaN  
...                     ...                    ...  
34999                   NaN                    NaN  
35000                   NaN                    NaN  
35001                   NaN                    NaN  
35002                   NaN                    NaN  
35003                   NaN                    NaN  

[35004 rows x 7 columns]

In [16]:
data_train.isna().sum()

year                        0
category                    0
topic                       0
indicator                   0
data_value               3154
confidence_limit_low     3522
confidence_limit_high    3522
dtype: int64

## Xử lý giá trị null

In [17]:
data_train['data_value'].fillna(data_train['data_value'].mean(), inplace = True)
data_train['confidence_limit_low'].fillna(data_train['confidence_limit_low'].mean(), inplace = True)
data_train['confidence_limit_high'].fillna(data_train['confidence_limit_high'].mean(), inplace = True)
data_train

/tmp/ipykernel_33/2894356728.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  data_train['data_value'].fillna(data_train['data_value'].mean(), inplace = True)
/tmp/ipykernel_33/2894356728.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_train['data_value'].fillna(data_train['data_value'].mean(), inplace = True)
/tmp/ipykernel_33/289

year                 category                   topic  \
0      2011  Cardiovascular Diseases  Coronary Heart Disease   
1      2011  Cardiovascular Diseases  Coronary Heart Disease   
2      2011  Cardiovascular Diseases  Coronary Heart Disease   
3      2011  Cardiovascular Diseases  Coronary Heart Disease   
4      2011  Cardiovascular Diseases  Coronary Heart Disease   
...     ...                      ...                     ...   
34999  2011  Cardiovascular Diseases  Coronary Heart Disease   
35000  2011  Cardiovascular Diseases  Coronary Heart Disease   
35001  2011  Cardiovascular Diseases  Coronary Heart Disease   
35002  2011  Cardiovascular Diseases  Coronary Heart Disease   
35003  2011  Cardiovascular Diseases  Coronary Heart Disease   

                                               indicator  data_value  \
0      Prevalence of coronary heart disease among US ...   36.370628   
1      Prevalence of coronary heart disease among US ...   36.370628   
2      Prevalence of coronary heart disease among US ...   36.370628   
3      Prevalence of coronary heart disease among US ...   36.370628   
4      Prevalence of coronary heart disease among US ...   36.370628   
...                                                  ...         ...   
34999  Prevalence of coronary heart disease among US ...   36.370628   
35000  Prevalence of coronary heart disease among US ...   36.370628   
35001  Prevalence of coronary heart disease among US ...   36.370628   
35002  Prevalence of coronary heart disease among US ...   36.370628   
35003  Prevalence of coronary heart disease among US ...   36.370628   

       confidence_limit_low  confidence_limit_high  
0                 32.607573              40.178076  
1                 32.607573              40.178076  
2                 32.607573              40.178076  
3                 32.607573              40.178076  
4                 32.607573              40.178076  
...                     ...                    ...  
34999             32.607573              40.178076  
35000             32.607573              40.178076  
35001             32.607573              40.178076  
35002             32.607573              40.178076  
35003             32.607573              40.178076  

[35004 rows x 7 columns]

## Tiền xử lý dữ liệu

In [18]:
lebelE = LabelEncoder()
data_train['topic'] = lebelE.fit_transform(data_train[['topic']])

/opt/conda/lib/python3.10/site-packages/sklearn/preprocessing/_label.py:116: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/tmp/ipykernel_33/2995355987.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_train['topic'] = lebelE.fit_transform(data_train[['topic']])


In [19]:
data_train.head()

year                 category  topic  \
0  2011  Cardiovascular Diseases      2   
1  2011  Cardiovascular Diseases      2   
2  2011  Cardiovascular Diseases      2   
3  2011  Cardiovascular Diseases      2   
4  2011  Cardiovascular Diseases      2   

                                           indicator  data_value  \
0  Prevalence of coronary heart disease among US ...   36.370628   
1  Prevalence of coronary heart disease among US ...   36.370628   
2  Prevalence of coronary heart disease among US ...   36.370628   
3  Prevalence of coronary heart disease among US ...   36.370628   
4  Prevalence of coronary heart disease among US ...   36.370628   

   confidence_limit_low  confidence_limit_high  
0             32.607573              40.178076  
1             32.607573              40.178076  
2             32.607573              40.178076  
3             32.607573              40.178076  
4             32.607573              40.178076

# Chia tập dữ liệu

In [20]:
y = data_train['topic']
X = data_train.drop(columns='topic', axis=1)
X_train_full, X_val_full, y_train, y_val = train_test_split(X, y, test_size=0.2, random_state=42)

In [21]:
# Chỉ dữ lại các cột đã chọn
categorical_cols = [cname for cname in X_train_full.columns if X_train_full[cname].nunique() < 10 
                    and X_train_full[cname].dtype == "object"]
numerical_cols = [cname for cname in X_train_full.columns if 
                  X_train_full[cname].dtype in ['int64', 'float64']]
my_cols = categorical_cols + numerical_cols
X_train = X_train_full[my_cols].copy()
X_val = X_val_full[my_cols].copy()

## Đóng gói quá trình

In [22]:
# Tạo pipeline cho tiền xử lý

# Cho number
numerical_transformer = SimpleImputer(strategy='mean')

# Cho kiểu category
categorical_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='most_frequent')),
    ('onehot', OneHotEncoder(handle_unknown='ignore'))
])

# Bộ tiền xử lý
preprocessor = ColumnTransformer(
    transformers=[
        ('num', numerical_transformer, numerical_cols),
        ('cat', categorical_transformer, categorical_cols)
    ]
)

## Xây dụng model


In [26]:
random_model = RandomForestClassifier(random_state=42)

params = { 
    'criterion': ['entropy'],
}

model = GridSearchCV(estimator=random_model, param_grid=params,
                           cv = 5,
                           n_jobs=-1, scoring="accuracy")
my_pipeline = Pipeline(steps=[('preprocessor', preprocessor), ('model', model)]).fit(X_train, y_train)

In [24]:
X_train

category  year  data_value  confidence_limit_low  \
24075             Risk Factors  2013   37.600000             35.200000   
26108             Risk Factors  2011   43.300000             41.100000   
32477             Risk Factors  2011   82.900000             81.200000   
776    Cardiovascular Diseases  2011   36.370628             32.607573   
480    Cardiovascular Diseases  2013   36.370628             32.607573   
...                        ...   ...         ...                   ...   
16850             Risk Factors  2011   27.900000             26.200000   
6265              Risk Factors  2013    7.900000              6.700000   
11284             Risk Factors  2014   19.700000             11.400000   
860    Cardiovascular Diseases  2011   36.370628             32.607573   
15795             Risk Factors  2013   26.700000             25.800000   

       confidence_limit_high  
24075              40.000000  
26108              45.500000  
32477              84.500000  
776                40.178076  
480                40.178076  
...                      ...  
16850              29.700000  
6265                9.400000  
11284              32.100000  
860                40.178076  
15795              27.600000  

[28003 rows x 5 columns]

## Predict Model

In [25]:
preds = my_pipeline.predict(X_val)
score = mean_absolute_error(y_val, preds)
print('MAE:', score)
accuracy_score(y_val, preds)

MAE: 1.1071275532066847


0.6667618911584059

## Không tham số: Entropy
- MAE: 1.1071275532066847
- 0.6667618911584059
## Với  'max_depth': range(3,16)
- MAE: 1.2583916583345236
- 0.6170547064705042
## Với 'max_depth': range(3,20)
- MAE: 1.1302671046993287
- 0.6597628910155692
## Với 'max_depth': range(3,50),
- MAE: 1.1139837166119126
- 0.6636194829310098
## Voi  'max_leaf_nodes': range(3,100)
- MAE: xxxx
- 0.533619482931009

## Voi 'max_depth': range(3,100), 'max_leaf_nodes': range(3,100)
## Voi 'max_features': list(range(0,20))
- MAE: 1.1185544922153978
- 0.6643336666190545

## Voi     'criterion': ['entropy', 'gini'], 'n_estimators': [25, 50, 100, 150, 200], 'max_features': ['sqrt', 'log2', None], 'max_depth': [3, 6, 9],  'max_leaf_nodes': range(3,30)
- MAE: 1.4185544922153978
- 0.43336666190545
